![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 3-MODUL. MA’LUMOTLARGA ISHLOV BERISH

### 3.2 Ma'lumotlarni tozalash va tayyorlash

## Sanalar bilan ishlash

In [1]:
import pandas as pd
import numpy as np

Pythonda sanalar bilan ishlash uchun maxsus `datetime` kutubxonasi mavjud.

In [2]:
import datetime as dt

Biz bu kutubxona bilan [Dasturlash asoslari kursimizda](https://python.sariq.dev/last-words/38-python-library#datetime-sana-va-vaqt) batafsil gaplashganimiz sababi ushbu darsimizda bevosita jadvallar va ulardagi sanalarni to'g'ri o'qishni o'rganamiz.

Keling darsimiz uchun kerakli ma'lumotlarni yuklab olamiz. Quyidagi jadvalda 1965-2016 yillar oralig'idagi zilzilalar haqida ma'lumotlar jamlangan.

In [3]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/blob/main/global_earthquake.csv?raw=true")
df.head()

,Date,Time,Latitude,Longitude,Type,Depth,Depth Error,Depth Seismic Stations,Magnitude,Magnitude Type,...,Magnitude Seismic Stations,Azimuthal Gap,Horizontal Distance,Horizontal Error,Root Mean Square,ID,Source,Location Source,Magnitude Source,Status
0,01/02/1965,13:44:18,19.246,145.616,Earthquake,131.6,NaN,NaN,6.0,MW,...,NaN,NaN,NaN,NaN,NaN,ISCGEM860706,ISCGEM,ISCGEM,ISCGEM,Automatic
1,01/04/1965,11:29:49,1.863,127.352,Earthquake,80.0,NaN,NaN,5.8,MW,...,NaN,NaN,NaN,NaN,NaN,ISCGEM860737,ISCGEM,ISCGEM,ISCGEM,Automatic
2,01/05/1965,18:05:58,-20.579,-173.972,Earthquake,20.0,NaN,NaN,6.2,MW,...,NaN,NaN,NaN,NaN,NaN,ISCGEM860762,ISCGEM,ISCGEM,ISCGEM,Automatic
3,01/08/1965,18:49:43,-59.076,-23.557,Earthquake,15.0,NaN,NaN,5.8,MW,...,NaN,NaN,NaN,NaN,NaN,ISCGEM860856,ISCGEM,ISCGEM,ISCGEM,Automatic
4,01/09/1965,13:32:50,11.938,126.427,Earthquake,15.0,NaN,NaN,5.8,MW,...,NaN,NaN,NaN,NaN,NaN,ISCGEM860890,ISCGEM,ISCGEM,ISCGEM,Automatic


Jadvalning `Date` ustuni zilzila sanasi,  `Time` ustuni esa zilzila vaqtini saqlaydi. Keling bu ustunlardagi ma'lumotlar turini ko'ramiz:

In [4]:
df[['Date','Time']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23409 entries, 0 to 23408
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    23409 non-null  object
 1   Time    23409 non-null  object
dtypes: object(2)
memory usage: 365.9+ KB


Ahamiyat bering, ikkala ustunning turini `object` deb ko'rsatayapti, `pandas` tilida bu matn degan, vaholanki, `pandas`da sanalar uchun maxsus `datetime64` turi bor.

Keling, `Date` ustunini alohida chiqaramiz:

In [5]:
df[['Date']].head()

,Date
0,01/02/1965
1,01/04/1965
2,01/05/1965
3,01/08/1965
4,01/09/1965


Keling `Date` ustun qiymatlarini to'g'ri ma'lumot turiga aylantirish uchun, avvalo qiymatlarning yozilishiga ahamiyat beramiz.

Jadval qiymatlarini `datetime`ga o'tkazish uchun `.to_datetime()` metodiga murojaat qilamiz, bunda biz jadvaldagi sanalr formatini aniq ko'rsatishimiz kerak.

Jadval qiymatlari `01/02/1965` ya'ni `oy/kun/yil` ko'rinishida yozilgan. `oy` va `kun` 2 xonali son, `yil` esa 4 xona ko'rinishida berilgan. Demak biz ham qiymatlarni shu ko'rinishda `datetime` formatiga o'tkazishimiz kerak.

**Diqqat! Ahamiyat bering, biz O'zbek tilida sanalarni `kun-oy-yil` tartibida yozishga o'rganganmiz, AQShda esa `oy-kun-yil` tartibi qabul qilingan.**

Oylarni `%m`, kunlarni `%d`, 4 xonali yilni `%Y` (2 xonali yillar uchun `%y`) deb ko'rsatamiz.

In [6]:
pd.to_datetime(df['Date'], format="%m/%d/%Y")

,Date
0,1965-01-02
1,1965-01-04
2,1965-01-05
3,1965-01-08
4,1965-01-09
...,...
23404,2016-12-28
23405,2016-12-28
23406,2016-12-28
23407,2016-12-29


Ko'rib turganingizdek, qiymatlar `datetime64` formatiga o'tdi. Biz bu qiymatlarni alohida ustunga saqlab olishimiz ham mumkin:

In [7]:
df['datetime'] = pd.to_datetime(df['Date'], format="%m/%d/%Y")
df[['datetime']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23409 entries, 0 to 23408
Data columns (total 1 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   datetime  23409 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 183.0 KB


Huddi shu kabi, `Time` ustuniga ham ishlov beramiz:

In [8]:
df[['Time']].head()

,Time
0,13:44:18
1,11:29:49
2,18:05:58
3,18:49:43
4,13:32:50


In [9]:
pd.to_datetime(df['Time'], format="%H:%M:%S")

,Time
0,1900-01-01 13:44:18
1,1900-01-01 11:29:49
2,1900-01-01 18:05:58
3,1900-01-01 18:49:43
4,1900-01-01 13:32:50
...,...
23404,1900-01-01 08:22:12
23405,1900-01-01 09:13:47
23406,1900-01-01 12:38:51
23407,1900-01-01 22:30:19


Bu safar vaqt bilan birga standart sana `1900-01-01` ham yozilib qoldi. Faqatgina vaqtni filtrlab olish uchun `dt.time` parametrini chaqiramiz.

`format` parametrini to'g'ri yozish uchun quyidagi jadvladan foydalaning: https://strftime.org/

In [10]:
pd.to_datetime(df['Time'], format="%H:%M:%S").dt.time

,Time
0,13:44:18
1,11:29:49
2,18:05:58
3,18:49:43
4,13:32:50
...,...
23404,08:22:12
23405,09:13:47
23406,12:38:51
23407,22:30:19


Saqlab olamiz

In [11]:
df['Time'] = pd.to_datetime(df['Time'], format="%H:%M:%S").dt.time

Biror qiymatni chaqirib ko'ramiz:

In [12]:
df.at[0,'Time']

datetime.time(13, 44, 18)

Istasak vaqtdan soatlarni yoki minutlarni ham ajratib olishimiz mumkin:

In [13]:
df.at[0,'Time'].hour

13

In [14]:
df.at[0,'datetime']

Timestamp('1965-01-02 00:00:00')

Yuqoridagi o'zgarishlardan so'ng `datetime` fromatidagi qiymatlar bilan ishlash ishimizni ancha osonlashtiradi. Misol uchun, biz zilzilalar qaysi oylarda bo'lganini ajratib olishimiz mumkin:

In [15]:
df['datetime'].dt.month

,datetime
0,1
1,1
2,1
3,1
4,1
...,...
23404,12
23405,12
23406,12
23407,12


Qiymatlarni alohiuda ustunga saqlab olib, ma'lumotlar tahllida ham ishlatish mumkin:

In [16]:
df['Month'] = df['datetime'].dt.month

Keling qaysi oylarda eng ko'p zilzila bo'lishini kuzatilganini ko'ramiz:

In [17]:
df['Month'].value_counts()

,count
Month,
3,2113
8,2014
12,2001
11,1987
9,1985
4,1970
5,1964
10,1952
1,1891


Yuqoridan eng ko'p zilzilalar Mart oyida (`2113` ta) eng kami esa Iyun oyida (`1824` ta) kuzatilganini ko'rishimiz mumkin.

### Savol: Zilzilalarning aksari haftaning qaysi kuniga to'g'ri kelishini chiqaring

In [18]:
df['datetime'].dt.weekday.value_counts()

,count
datetime,
5,3433
2,3431
4,3362
0,3349
3,3306
6,3284
1,3244
